libvast/vast/concept/parseable/core/rule.hpp

```diff
@@ -71,25 +165,24 @@ class rule : public parser<rule<Iterator, Attribute>> {
   rule() : parser_{std::make_shared<rule_pointer>()} {
   }
 
-  template <
-    class RHS,
-    class = std::enable_if_t<
-      is_parser_v<std::decay_t<RHS>> && !detail::is_same_or_derived_v<rule, RHS>
-    >
-  >
-  rule(RHS&& rhs)
-    : rule{} {
+  template <class RHS, class = std::enable_if_t<
+                         is_parser_v<std::decay_t<
+                           RHS>> && !detail::is_same_or_derived_v<rule, RHS>>>
+  rule(RHS&& rhs) : rule{} {
     make_parser<RHS>(std::forward<RHS>(rhs));
   }
 
   template <class RHS>
   auto operator=(RHS&& rhs)
-    -> std::enable_if_t<is_parser_v<std::decay_t<RHS>>
-                        && !detail::is_same_or_derived_v<rule, RHS>> {
+    -> std::enable_if_t<is_parser_v<std::decay_t<
+                          RHS>> && !detail::is_same_or_derived_v<rule, RHS>> {
     make_parser<RHS>(std::forward<RHS>(rhs));
   }
 
-  bool parse(Iterator& f, const Iterator& l, unused_type) const {
+  // This overload must not be available if Attribute == unused_type.
+  template <class Instantiate_on_use = bool>
+  std::enable_if_t<!std::is_same_v<Attribute, unused_type>, Instantiate_on_use>
+  parse(Iterator& f, const Iterator& l, unused_type) const {
```

Comment:
Do we need the overloads in the first place? All of our parser do a dispatching on `unused` anyways:

```c++
template <class T>
bool parse(Iterator& f, const Iterator& l, T&& x) {
  VAST_ASSERT(*parser_ != nullptr);
  return (*parser_)->parse(f, l, std::forward<T>(x));
}
```

Reply:
I followed your suggestion and removed the extra overload for `unused_type`.